### ISSUE TITLE
Broker can not respond to client requests

### ISSUE THREAD

**Issue by massakam** [apache/pulsar#3630]
**Describe the bug**

One of the brokers in our Pulsar cluster suddenly hung up. The broker process was alive, but it seems that it could not respond to requests from clients.

At that time, the following errors occurred on the client side:

```
02:35:08.736 [pulsar-client-io-5-8] WARN  o.a.p.c.i.BinaryProtoLookupService   - [persistent://xxx/global/xxx/xxx-partition-7] failed to send lookup request : org.apache.pulsar.client.api.PulsarClientException$TimeoutException: 9139700 lookup request timedout after ms 30000
java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException$TimeoutException: 9139700 lookup request timedout after ms 30000
```

```
02:35:17.739 [pulsar-client-io-5-8] WARN  o.a.pulsar.common.api.PulsarHandler  - [[id: 0x678cbfd2, L:/xxx.xxx.xxx.xxx:36338 - R:f1c2-broker103.pulsar.xxx.yahoo.co.jp/xxx.xxx.xxx.xxx:6651]] Forcing connection to close after keep-alive timeout
02:35:17.748 [pulsar-client-io-5-8] WARN  o.a.pulsar.client.impl.ConsumerImpl  - [persistent://xxx/global/xxx/xxx][sub1] Failed to subscribe to topic on f1c2-broker103.pulsar.xxx.yahoo.co.jp/xxx.xxx.xxx.xxx:6651
```

```
02:35:33.951 [pulsar-client-io-5-8] WARN  o.a.pulsar.client.impl.ClientCnx     - Error during handshake
javax.net.ssl.SSLException: handshake timed out
        at io.netty.handler.ssl.SslHandler.handshake(...)(Unknown Source) ~[netty-all-4.1.22.Final.jar:4.1.22.Final]
02:35:33.952 [pulsar-client-io-5-8] WARN  o.a.p.client.impl.ConnectionPool     - [[id: 0x208e5fb2, L:/xxx.xxx.xxx.xxx:60874 ! R:f1c2-broker103.pulsar.xxx.yahoo.co.jp/xxx.xxx.xxx.xxx:6651]] Connection handshake failed: org.apache.pulsar.client.api.PulsarClientException: Connection already closed
02:35:33.952 [pulsar-client-io-5-8] WARN  o.a.p.client.impl.ConnectionHandler  - [persistent://xxx/global/xxx/xxx] [xxx-497-2225404] Error connecting to broker: org.apache.pulsar.client.api.PulsarClientException: Connection already closed
02:35:33.952 [pulsar-client-io-5-8] WARN  o.a.p.client.impl.ConnectionHandler  - [persistent://xxx/global/xxx/xxx] [xxx-497-2225404] Could not get connection to broker: org.apache.pulsar.client.api.PulsarClientException: Connection already closed -- Will try again in 0.188 s
```

That broker returned to normal after restarting, and the errors no longer occur on the client side. The load of that broker was not high, so I think that there is a bug in broker code.

**Additional context**

Broker OS: CentOS Linux release 7.6.1810
Broker version: 2.2.1
Broker spec: Real server / 2.10GHz / 2CPU / 256GBMEM / SATA SSD 240GB x1 / 10G Base-T*2port

**Comment by massakam** [comment]
This is a thread dump acquired just before restarting the broker:
[threaddump.txt](https://github.com/apache/pulsar/files/2883472/threaddump.txt)

It seems that this thread is locking a `PersistentSubscription` object and the other pulsar-io threads are waiting for the lock to be released.
```
"pulsar-io-21-61" #1012 prio=5 os_prio=0 tid=0x00007f5cb006f000 nid=0x2627 waiting on condition [0x00007f5acd0e9000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00007f75cbe5a950> (a java.util.concurrent.CompletableFuture$Signaller)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
        at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
        at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at org.apache.pulsar.zookeeper.ZooKeeperDataCache.get(ZooKeeperDataCache.java:94)
        at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.isConsumersExceededOnTopic(PersistentDispatcherMultipleConsumers.java:148)
        at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.addConsumer(PersistentDispatcherMultipleConsumers.java:129)
        - locked <0x00007f6384d84368> (a org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers)
        at org.apache.pulsar.broker.service.persistent.PersistentSubscription.addConsumer(PersistentSubscription.java:139)
        - locked <0x00007f6373bd6600> (a org.apache.pulsar.broker.service.persistent.PersistentSubscription)
        at org.apache.pulsar.broker.service.persistent.PersistentTopic.lambda$9(PersistentTopic.java:517)
        at org.apache.pulsar.broker.service.persistent.PersistentTopic$$Lambda$229/184274928.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
        at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
        at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
        at org.apache.pulsar.broker.service.persistent.PersistentTopic.subscribe(PersistentTopic.java:513)
        at org.apache.pulsar.broker.service.ServerCnx.lambda$22(ServerCnx.java:609)
        at org.apache.pulsar.broker.service.ServerCnx$$Lambda$203/2100949060.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
        at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
        at org.apache.pulsar.broker.service.ServerCnx.lambda$18(ServerCnx.java:592)
        at org.apache.pulsar.broker.service.ServerCnx$$Lambda$201/788613052.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
        at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
        at org.apache.pulsar.broker.service.ServerCnx.lambda$6(ServerCnx.java:550)
        at org.apache.pulsar.broker.service.ServerCnx$$Lambda$197/133793716.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
        at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
        at org.apache.pulsar.broker.service.ServerCnx.handleSubscribe(ServerCnx.java:539)
        at org.apache.pulsar.common.api.PulsarDecoder.channelRead(PulsarDecoder.java:202)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1389)
        at io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1171)
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1205)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1414)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:945)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:806)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:404)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:304)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by massakam** [comment]
The same problem occurred on another broker server.
This is a thread dump acquired at that time:
[threaddump2.txt](https://github.com/apache/pulsar/files/2886151/threaddump2.txt)

Does anyone know the cause?

**Comment by sijie** [comment]
@massakam 

I think this is quite similar to what @rdhabalia hit - #3633 and what @merlimat fixed by #3591. 

#3591 is released as part of 2.3.0. it might be worth double confirming if #3591 addressed the issue here.

**Comment by massakam** [comment]
Thank you @sijie.

> I think this is quite similar to what @rdhabalia hit - #3633

I think so too. With @rdhabalia's modification, broker will be able to return to normal even if this deadlock occurs. However, I think it is ideal that this deadlock does not occur.

> #3591 is released as part of 2.3.0. it might be worth double confirming if #3591 addressed the issue here.

We have already backported #3591 to v2.2.1 and are using it. So this problem is not fixed by #3591.

**Comment by rdhabalia** [comment]
> even if this deadlock occurs. However, I think it is ideal that this deadlock does not occur.

**Comment by rdhabalia** [comment]
> even if this deadlock occurs. However, I think it is ideal that this deadlock does not occur.

@massakam, we need to figure out the root cause the issue so, it won't happen again. However, everytime I see deadlock, thread-dump showed that thread gets blocked while fetching data from zk and it happened multiple times at broker so, i don't believe that zk-client is failing this many times to return callback. However, I don't see any exception at that time which can give us evident to understand this issue. So, I feel there has to be an issue at pulsar app level rather zk-client level and I am looking into it to figure out actual root cause.

**Comment by massakam** [comment]
Each time a broker hung, one thread named `ForkJoinPool.commonPool-worker-x` always stopped in the following state:
```
"ForkJoinPool.commonPool-worker-30" #1098 daemon prio=5 os_prio=0 tid=0x00007f6104002800 nid=0x267f waiting for monitor entry [0x00007f5a85b51000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.apache.pulsar.broker.service.persistent.PersistentSubscription.addConsumer(PersistentSubscription.java:100)
	- waiting to lock <0x00007f6373bd6600> (a org.apache.pulsar.broker.service.persistent.PersistentSubscription)
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.lambda$9(PersistentTopic.java:517)
	at org.apache.pulsar.broker.service.persistent.PersistentTopic$$Lambda$229/184274928.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.subscribe(PersistentTopic.java:513)
	at org.apache.pulsar.broker.service.ServerCnx.lambda$22(ServerCnx.java:609)
	at org.apache.pulsar.broker.service.ServerCnx$$Lambda$203/2100949060.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at org.apache.pulsar.broker.service.ServerCnx.lambda$18(ServerCnx.java:592)
	at org.apache.pulsar.broker.service.ServerCnx$$Lambda$201/788613052.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$null$0(PulsarAuthorizationProvider.java:129)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider$$Lambda$199/1961028585.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$canConsumeAsync$1(PulsarAuthorizationProvider.java:128)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider$$Lambda$198/467223006.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.apache.pulsar.zookeeper.ZooKeeperDataCache.lambda$0(ZooKeeperDataCache.java:66)
	at org.apache.pulsar.zookeeper.ZooKeeperDataCache$$Lambda$25/41765385.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.apache.pulsar.zookeeper.ZooKeeperCache.lambda$7(ZooKeeperCache.java:347)
	at org.apache.pulsar.zookeeper.ZooKeeperCache$$Lambda$23/778720569.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.apache.pulsar.zookeeper.ZooKeeperCache.lambda$11(ZooKeeperCache.java:328)
	at org.apache.pulsar.zookeeper.ZooKeeperCache$$Lambda$27/1695511770.run(Unknown Source)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```
This thread seems to be related to the processing of ZK cache. Could this be the cause of this problem?

**Comment by massakam** [comment]
- This problem seems likely to occur in the following situations:
  - Consumer creation and closing are repeated many times in parallel using the WebSocket API
  - The topic is partitioned
  - The subscription type is Shared
- This problem no longer occurs (or has become difficult to occur) when I revert https://github.com/apache/pulsar/pull/2785

**Comment by massakam** [comment]
I wrote a simple test code of caffeine used for ZK cache:
https://gist.github.com/massakam/3e15f0bad0260b17382dbc4bc785aa25#file-caffeinetest-java

Executing this code often results in the following:
https://gist.github.com/massakam/3e15f0bad0260b17382dbc4bc785aa25#file-result-sh

In this case, `ForkJoinPool.commonPool-worker-3` slept for 10 seconds.

> 22:03:10.183 [0] [ForkJoinPool.commonPool-worker-3] [INFO] Sleeping 10 sec...

Then, these three processes were completed about 10 seconds after the other processes were completed. Even though there were five threads named `ForkJoinPool.commonPool-worker-x`, threads other than `ForkJoinPool.commonPool-worker-3` were were idle.

> 22:03:20.186 [3] [ForkJoinPool.commonPool-worker-3] [INFO] key: 6, value: val6
> 22:03:20.186 [9] [ForkJoinPool.commonPool-worker-3] [INFO] key: 6, value: val6
> 22:03:20.186 [1] [ForkJoinPool.commonPool-worker-3] [INFO] key: 6, value: val6

I think that the same thing as this is happening in broker.

**Comment by massakam** [comment]
1. A thread `pulsar-io-x-x` locks a `PersistentSubscription` object to add a consumer.
2. In order to add the consumer, it is necessary to refer to `max_consumers_per_topic` included in the policy of the namespace, and it is assumed that a thread `ForkJoinPool.commonPool-worker-x` is in charge of that.
3. However, suppose that thread `ForkJoinPool.commonPool-worker-x` tries to add another consumer to the same subscription of the same topic before the process 2.
4. 3 does not complete as the `PersistentSubscription` object is locked. 2 also does not complete as the thread `ForkJoinPool.commonPool-worker-x` remains blocked.

What do you think about the above assumption?

**Comment by rdhabalia** [comment]
> Consumer creation and closing are repeated many times in parallel using the WebSocket API

yes, this happens when cache gets expired (after 5 mins) and at the same time many, threads tries to access the same key from the cache (so, it mainly occurs while adding-consumer which requires to to access global-zk cache). In our case, one of the client was keep disconnecting and creating new consumers and broker was going into deadlock state with in 10-20 mins.

> also does not complete as the thread `ForkJoinPool.commonPool-worker-x` remains blocked.

Hmm..ok, so, should we process zk-cache callback into separate external executor rather using default `ForkJoinPool` thread executor that also used by cache internally.?

I have created PR #3708 with the change so, will you be able to verify if this patch fixes the issue.

**Comment by rdhabalia** [comment]
@massakam 
I think you are right. thread-dump shows the same deadlock condition where
1. `pulsar-io` takes lock on `PersistentSubscription's` synchronized method and waiting for zk response
2. `ForkJoinPool` gets value from zk and same thread tries to access same lock taken by `pulsar-io` thread in `PersistentSubscription`. so, `ForkJoinPool-thread` couldn't complete and can't serve further zk-cache request. and the same time pulsar-io threads are waiting for zk-cache respinse.


```
ForkJoinPool.commonPool-worker-19" #381 daemon prio=5 os_prio=0 tid=0x00002af1fc005800 nid=0x2a4b waiting for monitor entry [0x00002af017202000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.apache.pulsar.broker.service.persistent.PersistentSubscription.addConsumer(PersistentSubscription.java:100)
	- waiting to lock <0x000000058e35a2b0> (a org.apache.pulsar.broker.service.persistent.PersistentSubscription)
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.lambda$9(PersistentTopic.java:523)
	at org.apache.pulsar.broker.service.persistent.PersistentTopic$$Lambda$309/503840967.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.subscribe(PersistentTopic.java:519)
	at org.apache.pulsar.broker.service.ServerCnx.lambda$22(ServerCnx.java:609)
	at org.apache.pulsar.broker.service.ServerCnx$$Lambda$299/1090387422.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at org.apache.pulsar.broker.service.ServerCnx.lambda$18(ServerCnx.java:592)
	at org.apache.pulsar.broker.service.ServerCnx$$Lambda$298/1756865666.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$null$2(PulsarAuthorizationProvider.java:129)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider$$Lambda$296/297063554.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider.lambda$canConsumeAsync$3(PulsarAuthorizationProvider.java:128)
	at org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider$$Lambda$295/1368029516.accept(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.apache.pulsar.zookeeper.ZooKeeperDataCache.lambda$0(ZooKeeperDataCache.java:66)
	at org.apache.pulsar.zookeeper.ZooKeeperDataCache$$Lambda$25/992743048.accept(Unknown Source)
:
:
```

```
"pulsar-io-21-15" #340 prio=5 os_prio=0 tid=0x00002af27c020800 nid=0x263d waiting on condition [0x00002aefdfb09000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00000006960b8018> (a java.util.concurrent.CompletableFuture$Signaller)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
        at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
        at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at org.apache.pulsar.zookeeper.ZooKeeperDataCache.get(ZooKeeperDataCache.java:94)
        at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.isConsumersExceededOnTopic(PersistentDispatcherMultipleConsumers.java:148)
        at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.addConsumer(PersistentDispatcherMultipleConsumers.java:129)
        - locked <0x000000058afd2dd0> (a org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers)
        at org.apache.pulsar.broker.service.persistent.PersistentSubscription.addConsumer(PersistentSubscription.java:139)
        - locked <0x000000058afd2ed8> (a org.apache.pulsar.broker.service.persistent.PersistentSubscription)
        at org.apache.pulsar.broker.service.persistent.PersistentTopic.lambda$9(PersistentTopic.java:523)
        at org.apache.pulsar.broker.service.persistent.PersistentTopic$$Lambda$309/503840967.accept(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
        at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
        at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
        at org.apache.pulsar.broker.service.persistent.PersistentTopic.subscribe(PersistentTopic.java:519)
        at org.apache.pulsar.broker.service.ServerCnx.lambda$22(ServerCnx.java:609)
        at org.apache.pulsar.broker.service.ServerCnx$$Lambda$299/1090387422.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
        at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
        at org.apache.pulsar.broker.service.ServerCnx.lambda$18(ServerCnx.java:592)
        at org.apache.pulsar.broker.service.ServerCnx$$Lambda$298/1756865666.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
        at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
        at org.apache.pulsar.broker.service.ServerCnx.lambda$6(ServerCnx.java:550)
        at org.apache.pulsar.broker.service.ServerCnx$$Lambda$294/380312495.apply(Unknown Source
```

**Comment by massakam** [comment]
@rdhabalia I modified the test code `CaffeineTest.java` of comment https://github.com/apache/pulsar/issues/3630#issuecomment-467926959 as follows, but the same phenomenon occurred. Just using an external executor may not solve this problem.

Change:
```diff
--- CaffeineTest.java.BAK       2019-02-28 10:56:31.401268202 +0900
+++ CaffeineTest.java   2019-02-28 11:45:04.092022888 +0900
@@ -7,19 +7,23 @@
 import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
 import com.github.benmanes.caffeine.cache.Caffeine;

+import org.apache.bookkeeper.common.util.OrderedExecutor;
+
 public class CaffeineTest {
     private final AsyncLoadingCache<String, String> dataCache;
+    private final OrderedExecutor executor;

     public CaffeineTest() {
         dataCache = Caffeine.newBuilder().expireAfterWrite(1, TimeUnit.MINUTES).buildAsync((k, executor) -> null);
+        executor = OrderedExecutor.newBuilder().name("external-executor").numThreads(6).build();
     }

     public CompletableFuture<String> getDataAsync(final String key) {
         CompletableFuture<String> future = new CompletableFuture<>();

-        dataCache.get(key, (k, executor) -> {
+        dataCache.get(key, (k, cacheExecutor) -> {
             CompletableFuture<String> cacheFuture = new CompletableFuture<>();
-            executor.execute(() -> {
+            this.executor.execute(() -> {
                 cacheFuture.complete("val" + k);
             });
             return cacheFuture;
@@ -33,6 +37,10 @@
         return future;
     }

+    public void shutdown() {
+        executor.shutdown();
+    }
+
     public static void main(String[] args) throws Exception {
         System.out.println("---- START ----");
         CaffeineTest cTest = new CaffeineTest();
@@ -49,6 +57,7 @@
             threads[i].join();
         }
         System.out.println("---- FINISH ----");
+        cTest.shutdown();
     }
 }

@@ -74,7 +83,7 @@
                 final String key = String.valueOf(i);
                 cTest.getDataAsync(key).thenAccept(v -> {
                     printInfo(String.format("key: %s, value: %s", key, v));
-                    if (tid == 0 && Thread.currentThread().getName().startsWith("ForkJoinPool.commonPool-worker")
+                    if (tid == 0 && Thread.currentThread().getName().startsWith("external-executor-OrderedExecutor")
                             && !alreadySlept) {
                         alreadySlept = true;
                         printInfo("Sleeping 10 sec...");
```

Execution result:
```sh
---- START ----
11:16:18.454 [9] [Thread-10] [INFO] key: 0, value: val0
11:16:18.456 [6] [Thread-7] [INFO] key: 0, value: val0

...

11:16:18.472 [1] [Thread-2] [INFO] key: 3, value: val3
11:16:18.481 [7] [external-executor-OrderedExecutor-2-0] [INFO] key: 6, value: val6
11:16:18.482 [4] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
11:16:18.482 [0] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8

# "external-executor-OrderedExecutor-2-0" starts sleeping for 10 seconds here
11:16:18.483 [0] [external-executor-OrderedExecutor-2-0] [INFO] Sleeping 10 sec...

11:16:18.483 [1] [Thread-2] [INFO] key: 4, value: val4
11:16:18.483 [1] [Thread-2] [INFO] key: 5, value: val5
11:16:18.483 [1] [Thread-2] [INFO] key: 6, value: val6
11:16:18.484 [1] [Thread-2] [INFO] key: 7, value: val7
11:16:18.484 [1] [Thread-2] [INFO] key: 8, value: val8
11:16:18.484 [1] [Thread-2] [INFO] key: 9, value: val9

# Processing of all threads is stopped for about 10 seconds from here

11:16:28.483 [8] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
11:16:28.484 [2] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
11:16:28.484 [3] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
11:16:28.484 [9] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
11:16:28.485 [6] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
11:16:28.485 [5] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
11:16:28.485 [7] [external-executor-OrderedExecutor-2-0] [INFO] key: 8, value: val8
---- FINISH ----
```

**Comment by xbblq62** [comment]
We are hitting a similar issue where all operations performed on a broker gets timed out. 
Error in pulsar client:
Pulsar Handshake was not completed within timeout, closing connection

Have tried to run commands via pulsar-admin as well and face timeout exception for any commands ran on the broker.

In our case, we generate large number of topics dynamically on the fly. After a while, under peak load we start facing the issue.

@rdhabalia Please share if there are any work-arounds to avoid this issue?

**Comment by rdhabalia** [comment]
@xbblq62 this issue should be fixed by: https://github.com/apache/pulsar/pull/8304

**Comment by xbblq62** [comment]
Thanks @rdhabalia 
Have upgraded to 2.6.2 and still the issue occurs. May be the symptoms of issue is similar to the above mentioned but the problem might be different.

found the following errors in the broker which is down:

`13:10:32.240 [pulsar-2-1] ERROR org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl - Deleting cursor`

`13:11:14.571 [pulsar-zk-session-watcher-5-1] ERROR org.apache.pulsar.zookeeper.ZooKeeperSessionWatcher - timeout expired for reconnecting, invoking shutdown service`

`13:11:44.688 [pulsar-web-38-5] ERROR org.apache.pulsar.broker.web.PulsarWebResource - Finding owner for topic xxx timed out`

`13:39:46.036 [pulsar-io-21-7] WARN  org.apache.pulsar.broker.service.ServerCnx - [/ip:port] Got exception io.netty.channel.unix.Errors$NativeIoException: readAddress(..) failed: Connection reset by peer`

**Comment by rdhabalia** [comment]
@wolfstudy I think #8304 is not correctly cherry-picked into 2.6 branch.  we have removed data-cache expiry on [ZooKeeperCache.java#L112](https://github.com/apache/pulsar/blob/branch-2.6/pulsar-zookeeper-utils/src/main/java/org/apache/pulsar/zookeeper/ZooKeeperCache.java#L112) but it still there in 2.6 branch. can we fix it?

@xbblq62  Is it possible for you to try latest master changes to pick up #8304 change properly.

**Comment by sijie** [comment]
@wolfstudy @rdhabalia let's create a fix on branch-2.6 and create a 2.6.3 release.

**Comment by eolivelli** [comment]
@nicoloboschi FYI

**Comment by lhotari** [comment]
> let's create a fix on branch-2.6 and create a 2.6.3 release.

@sijie @wolfstudy @rdhabalia for consistency, it might be good to also cherry pick d8f319ea , #8302 before removing the extra line (expireAfterWrite from dataCache) to fix the cherry-picking of #8304. There wouldn't have been a merge conflict if that change would have been included.

**Comment by hadesy** [comment]
> @wolfstudy @rdhabalia let's create a fix on branch-2.6 and create a 2.6.3 release.

@sijie any update? we had the same problem on 2.6.2
```
[ForkJoinPool.commonPool-worker-3] ERROR org.apache.pulsar.broker.service.persistent.PersistentTopic - [persistent://cdc/mysql/test-partition-0] Policies update failed org.apache.pulsar.broker.service.BrokerServiceException$ServerMetadataException: java.util.concurrent.TimeoutException, scheduled retry in 60 seconds
java.util.concurrent.CompletionException: org.apache.pulsar.broker.service.BrokerServiceException$ServerMetadataException: java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:661) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:683) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2010) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.checkReplicationAndRetryOnFailure(PersistentTopic.java:1027) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.onPoliciesUpdate(PersistentTopic.java:1844) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	at org.apache.pulsar.broker.service.BrokerService.lambda$null$59(BrokerService.java:1537) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	at java.util.Optional.ifPresent(Optional.java:159) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.service.BrokerService.lambda$null$60(BrokerService.java:1537) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:683) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2010) ~[?:1.8.0_252]
	at org.apache.pulsar.broker.service.BrokerService.lambda$onUpdate$61(BrokerService.java:1532) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	at org.apache.pulsar.common.util.collections.ConcurrentOpenHashMap$Section.forEach(ConcurrentOpenHashMap.java:388) ~[org.apache.pulsar-pulsar-common-2.6.2.jar:2.6.2]
	at org.apache.pulsar.common.util.collections.ConcurrentOpenHashMap.forEach(ConcurrentOpenHashMap.java:160) ~[org.apache.pulsar-pulsar-common-2.6.2.jar:2.6.2]
	at org.apache.pulsar.broker.service.BrokerService.onUpdate(BrokerService.java:1528) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	at org.apache.pulsar.broker.service.BrokerService.onUpdate(BrokerService.java:171) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	at org.apache.pulsar.zookeeper.ZooKeeperDataCache.lambda$4(ZooKeeperDataCache.java:138) ~[org.apache.pulsar-pulsar-zookeeper-utils-2.6.2.jar:2.6.2]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975) ~[?:1.8.0_252]
	at org.apache.pulsar.zookeeper.ZooKeeperCache.lambda$18(ZooKeeperCache.java:386) ~[org.apache.pulsar-pulsar-zookeeper-utils-2.6.2.jar:2.6.2]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975) ~[?:1.8.0_252]
	at org.apache.pulsar.zookeeper.ZooKeeperCache.lambda$14(ZooKeeperCache.java:366) ~[org.apache.pulsar-pulsar-zookeeper-utils-2.6.2.jar:2.6.2]
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402) [?:1.8.0_252]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) [?:1.8.0_252]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) [?:1.8.0_252]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) [?:1.8.0_252]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) [?:1.8.0_252]
Caused by: org.apache.pulsar.broker.service.BrokerServiceException$ServerMetadataException: java.util.concurrent.TimeoutException
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.checkReplication(PersistentTopic.java:1081) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	... 29 more
Caused by: java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1784) ~[?:1.8.0_252]
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928) ~[?:1.8.0_252]
	at org.apache.pulsar.zookeeper.ZooKeeperDataCache.get(ZooKeeperDataCache.java:97) ~[org.apache.pulsar-pulsar-zookeeper-utils-2.6.2.jar:2.6.2]
	at org.apache.pulsar.broker.service.persistent.PersistentTopic.checkReplication(PersistentTopic.java:1077) ~[org.apache.pulsar-pulsar-broker-2.6.2.jar:2.6.2]
	... 29 more
```

**Comment by sijie** [comment]
I think @wolfstudy and @codelipenghui are cherry-picking that change. we will see if we can cut a 2.6.3 release next week.

**Comment by jiazhai** [comment]
> > let's create a fix on branch-2.6 and create a 2.6.3 release.
> 
> @sijie @wolfstudy @rdhabalia for consistency, it might be good to also cherry pick [d8f319e](https://github.com/apache/pulsar/commit/d8f319ea1916589dcd2af858b4b995dc89b03840) , #8302 before removing the extra line (expireAfterWrite from dataCache) to fix the cherry-picking of #8304. There wouldn't have been a merge conflict if that change would have been included.
> 
> It seems that proper test coverage for ZooKeeperCache class is missing. Would it be possible to improve the test coverage?

Thanks @lhotari for the information.  Yes, we need to add tests to cover this.

**Comment by frankjkelly** [comment]
@wolfstudy / @codelipenghui can you tell which is the earliest release that has the fix mentioned here https://github.com/apache/pulsar/issues/3630#issuecomment-728721439
Thanks!

**Comment by kwenZh** [comment]
We are hitting a similar issue , Proxy stops working for proxying Broker connections while Admin API proxying keeps working. but directly connect pulsar-broker is normally,  proxy back to normal  when i restart pulsar-proxy

The proxy logs are filled with this type of warnings:
```
13:13:09.996 [pulsar-proxy-io-2-1] WARN org.apache.pulsar.common.protocol.PulsarHandler - [[id: 0x83e12747, L:/ip:port - R:/ip：port]] Pulsar Handshake was not completed within timeout, closing connection
```

**Comment by tisonkun** [comment]
Closed as stale. Please create a new issue if it's still relevant to the maintained versions.

@kwenzh If you're using a maintained version, for better visibility, I suggest you open a new issue if you're sure that it's a bug or ask in the [Q&A forum](https://github.com/apache/pulsar/discussions/categories/q-a) for help first.
